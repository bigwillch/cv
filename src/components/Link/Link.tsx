import React, { useContext } from 'react';

import Obfuscate from 'react-obfuscate';
import classNames from 'classnames';

import { SideLoaderContext } from 'Contexts';

declare const ENV: Array<string>;

type Props = {
  children: string;
  href: string;
};

export const Link: React.FC<Props> = ({
  children,
  href,
}) => {
  const {
    actions: {
      sideTrigger,
    },
  } = useContext(SideLoaderContext);

  // contact details need to be set as env vars
  // CONTACTEMAIL
  // CONTACTTEL
  switch (children) {
    case 'email':
    case 'tel':
      if (!ENV[href]) {
        return null;
      }
      return (
        <Obfuscate
          className={children}
          {...{ [children]: ENV[href] }}
        />
      );
    default:
  }

  // check href to check for example hashtag
  const hrefSplit = href.split('#');
  const sideload = hrefSplit[1] === 'example';
  const text = sideload ? 'Load example' : children;
  const className = classNames(
    sideload && 'button button--forward',
  );

  return (
    <a
      href={hrefSplit[0]}
      target="_blank"
      rel="noreferrer"
      className={className}
      // if preview hashtag present trigger sideload action
      onClick={sideload ? (e) => {
        e.preventDefault();
        sideTrigger(
          hrefSplit[0],
          { desc: children },
        );
      } : null}
    >
      { text }
    </a>
  );
};
