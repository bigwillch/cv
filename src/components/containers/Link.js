import React from 'react'
import { connect } from 'react-redux';
import { sideTrigger } from 'Redux/actions/sideloader'

import Obfuscate from 'react-obfuscate'
import classNames from 'classnames'

const mapDispatchToProps = (dispatch) => {
  return {
    sideTrigger: (href, data) => {
      dispatch(sideTrigger(href, data))
    }
  }
}

let Link = (props) => {

  switch (props.children) {
    case 'email':
    case 'tel':
      if (!ENV[props.href]) {
        return null
      } 
      const componentProps = {};
      componentProps[props.children] = ENV[props.href]
      return (
        <Obfuscate className={ props.children } { ...componentProps } />
      )
  }

  // check href to check for example hashtag
  const href = props.href.split('#'),
        sideload = href[1] === 'example' ? true : false,
        text = sideload ? 'Load example' : props.children,
        className = classNames(
          sideload && 'button button--forward'
        )

  return (
    <a 
      href={ href[0] }
      target="_blank"
      className={className}
      // if preview hashtag present trigger sideload action
      onClick={sideload ? (e) => {
        e.preventDefault()
        props.sideTrigger(
          href[0],
          { desc: props.children }
        )
      } : null } 
    >
      { text }
    </a>
  )
}

export default Link = connect(
  null,
  mapDispatchToProps
)(Link)