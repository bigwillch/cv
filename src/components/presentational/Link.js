import React from 'react'
import Obfuscate from 'react-obfuscate'

const Link = (props) => {

  switch (props.children) {
    case 'email':
    case 'tel':
      if (!ENV[props.href]) {
        return null
      } 
      const componentProps = {};
      componentProps[props.children] = ENV[props.href]
      return (
        <Obfuscate className={ props.children } { ...componentProps } />
      )
  default:
    return (
      <a href={ props.href } target="_blank">{ props.children }</a>
    )
  }

};
export default Link;