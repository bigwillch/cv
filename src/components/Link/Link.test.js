import React from 'react';
import { Link } from './Link';
import { SideLoaderProvider } from 'Contexts';
import toJson from 'enzyme-to-json';
import Obfuscate from 'react-obfuscate';

const createLinkWrapper = (href, text = 'Basic Link') => mount(
  <SideLoaderProvider>
    <Link
      href={href}
    >
      { text }
    </Link>
  </SideLoaderProvider>
)

describe('Link', () => {

  it('should render external links', () => {

    const wrapper = createLinkWrapper('//www.google.com');

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('a').hasClass('button')).toEqual(false);
    expect(wrapper.find('a').hasClass('button--forward')).toEqual(false);
    expect(wrapper.find('a').prop('onClick')).toEqual(null);
    expect(wrapper.find('a').text()).toEqual('Basic Link');
  });

  it('should render hashtagged example links', () => {

    const wrapper = createLinkWrapper('/examples/page#example');

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('a').hasClass('button')).toEqual(true);
    expect(wrapper.find('a').hasClass('button--forward')).toEqual(true);
    expect(typeof wrapper.find('a').prop('onClick')).toEqual('function');
    expect(wrapper.find('a').text()).toEqual('Load example');
  });

  it('should render obfuscated email links', () => {

    const email = 'hidden@hidden.com';
    global.ENV.CONTACTEMAIL = email;

    const wrapper = createLinkWrapper('CONTACTEMAIL', 'email');

    expect(wrapper.find(Obfuscate).hasClass('email')).toEqual(true);
    expect(wrapper.find(Obfuscate).prop('email')).toEqual(email);

    delete global.ENV.CONTACTEMAIL;

  });

  it('should render obfuscated telephone links', () => {

    const tel = '0777777777';
    global.ENV.CONTACTTEL = tel;

    const wrapper = createLinkWrapper('CONTACTTEL', 'tel');

    expect(wrapper.find(Obfuscate).hasClass('tel')).toEqual(true);
    expect(wrapper.find(Obfuscate).prop('tel')).toEqual(tel);

    delete global.ENV.CONTACTTEL;

  });

});