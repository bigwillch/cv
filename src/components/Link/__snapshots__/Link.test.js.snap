// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render external links 1`] = `
<SideLoaderProvider>
  <Link
    href="//www.google.com"
  >
    <a
      className=""
      href="//www.google.com"
      onClick={null}
      rel="noreferrer"
      target="_blank"
    >
      Basic Link
    </a>
  </Link>
</SideLoaderProvider>
`;

exports[`Link should render hashtagged example links 1`] = `
<SideLoaderProvider>
  <Link
    href="/examples/page#example"
  >
    <a
      className="button button--forward"
      href="/examples/page"
      onClick={[Function]}
      rel="noreferrer"
      target="_blank"
    >
      Load example
    </a>
  </Link>
</SideLoaderProvider>
`;
